src/NCI.OCPL.Api.CTSListingPages/Services/ESListingInfoQueryService.cs

```diff
@@ -79,17 +80,94 @@ public async Task<ListingInfo> GetByPrettyUrlName(string prettyUrlName)
+        /// <returns>A ListingResults object or null if exact or partial matches are not found.</returns>
+        public async Task<ListingResults> GetByIds(string[] ccodes)
+        {
+            // Set up the SearchRequest to send to elasticsearch.
+            Indices index = Indices.Index(new string[] { this._apiOptions.ListingInfoAliasName });
+            Types types = Types.Type(new string[] { "ListingInfo" });
+            SearchRequest request = new SearchRequest(index, types)
+            {
+                Query = new TermsQuery { Field = "concept_id", Terms = ccodes }
+            };
+
+            ISearchResponse<ListingInfo> response = null;
+            try
+            {
+                response = await _elasticClient.SearchAsync<ListingInfo>(request);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, $"Error searching index: '{this._apiOptions.ListingInfoAliasName}'.");
+                throw new APIInternalException("errors occured");
+            }
+
+            if (!response.IsValid)
+            {
+                String msg = $"Invalid response when searching for c-code(s) '{String.Join(",", ccodes)}'.";
+                _logger.LogError(msg);
+                _logger.LogError(response.DebugInformation);
+                throw new APIInternalException("errors occured");
+            }
+
+            ListingResults results = null;
+
+            // If there is exactly one response, the lookup was successful.
+            if (response.Total == 1)
+            {
+                results = new ListingResults()
+                {
+                    Meta = new ResultsMetadata
+                    {
+                        TotalResults = Convert.ToInt32(response.Total),
+                        From = 0
+                    },
+                    Results = new ListingInfo[] { response.Documents.First() }
+                };
+            }
+            // If there are more than one record in the response, log a warning. The controller will throw an error.
+            if (response.Total > 1)
+            {
+                // Build the array of ListingInfo items for the returned results.
+                List<ListingInfo> listingInfoResults = new List<ListingInfo>();
+                foreach (ListingInfo res in response.Documents)
+                {
+                    listingInfoResults.Add(res);
+                }
+
+                results = new ListingResults()
+                {
+                    Meta = new ResultsMetadata
+                    {
+                        TotalResults = Convert.ToInt32(response.Total),
+                        From = 0
+                    },
+                    Results = listingInfoResults.ToArray()
+                };
+
+                // Log an error
+                String msg = $"Multiple records found when searching for c-code(s) '{String.Join(",", ccodes)}'.";
+                _logger.LogWarning(msg);
+            }
```

Comment:
Updated! With the change to the return object, the controller does the logic for error throwing, while the service just sets up the array of `ListingInfo` objects. The service does log a warning for multiple returned, just in case.

Reply:
A few things here.

One, you're still doing too much work here. Instead of looping through the response.Documents collection and individually allocating the `List<T>` items, you can simply pass the collection into the `List<T>` constructor.  e.g. `new List<ListingInfo>(response.Documents)`.

But, I don't believe there's a reason to go through an intermediate collection.  You're returning an array, so a long as you have a valid response, you should be able to simply `return response.Documents.ToArray();` (The extra allocations are an anti-pattern we've picked up because we're all copying code which was originally written by someone who was new to the .Net platform.)

Also, in the interest of full disclosure: My original feedback was partially incorrect.  If we were still working with the possibility of paged output, you would have needed to return the metadata object as the total number of records is only available from the elasticsearch return. But since we're no longer living in that world, having the query return everything it finds is the correct outcome.

Reply:
Updated to have a simpler return of the response.Documents.